casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -177,8 +183,54 @@ sealed abstract class MultiParentCasperInstances {
           tip       = estimates.head
           _ <- Log[F].info(
                 s"CASPER: New fork-choice tip is block ${PrettyPrinter.buildString(tip.blockHash)}.")
+          lastFinalizedBlock        <- lastFinalizedBlockContainer.get
+          updatedLastFinalizedBlock <- updateLastFinalizedBlock(dag, lastFinalizedBlock)
+          _                         <- lastFinalizedBlockContainer.set(updatedLastFinalizedBlock)
         } yield attempt
 
+      def updateLastFinalizedBlock(dag: BlockDag,
+                                   lastFinalizedBlock: BlockMessage): F[BlockMessage] = {
+        val maybeFinalizedBlockChildren = dag.childMap.get(lastFinalizedBlock.blockHash)
+        maybeFinalizedBlockChildren match {
```

Comment:
This should be able to be implemented in much less code, but `cats` is dumb and missing `findM`.
```scala
def updateLastFinalizedBlock(dag: BlockDag, lastFinalizedBlock: BlockMessage): F[BlockMessage] =
  for {
    childrenHashes <- dag.childMap.getOrElse(lastFinalizedBlock.blockHash, Set.empty[BlockHash]).pure[F]
    children <- childrenHashes.traverse(BlockStore[F].get(_))
    finalizedChild <- children.findM(block => {
      val ft = SafetyOracle[F].normalizedFaultTolerance(dag, block)
      ft > faultToleranceThreshold
    })
    newFinalizedBlock <- finalizedChild.fold(lastFinalizedBlock.pure[F])(updateLastFinalizedBlock(dag, _))
  } yield newFinalizedBlock 
```

Reply:
Done in https://github.com/rchain/rchain/pull/1257